app/src/androidTest/java/ca/marshallasch/veil/database/DatabaseTest.java

```diff
@@ -68,26 +80,34 @@ public void checkBlocked()
     @Test
     public void createUser()
     {
-        assertNotNull(database.createUser("Marshall", "Asch", "email@email.com", "abc123"));
-        assertNotNull(database.createUser("Marshall", "Asch", "email@email.com", "abc123"));
-        assertNotNull(database.createUser("Marshall", "Asch", "email2@email.com", "abc1234"));
+        assertNotNull(database.createUser("Marshall", "Asch", "email_01@email.com", "abc123"));
+
+        //a duplicate user is defined as a duplicate userID, not duplicate usernames
+        assertNotNull(database.createUser("Marshall", "Asch", "email_01@email.com", "abc123"));
+        assertNotNull(database.createUser("Marshall", "Asch", "email_03@email.com", "abc1234"));
     }
 
     @Test
     public void login()
     {
-        assertNotNull(database.login("email@email.com", "abc123"));
-        assertNotNull(database.login("email2@email.com", "abc1234"));
-        assertNull(database.login("email2@email.com", "abc123"));
-        assertNull(database.login("email46@email.com", "abc123"));
+        assertNotNull(database.createUser("userFname1", "userLname1", "email_04@email.com", "abc123"));
+
+        assertNotNull(database.login("email_04@email.com", "abc123"));
+        assertNull(database.login("email_04@email.com", "*"));      // check some injections
+        assertNull(database.login("email_04@email.com", "%"));      // check another one
+        assertNull(database.login("non_existant_email_01@email.com", "abc123"));
     }
 
     @Test
-    public void insertKnownPost(){
-        //null comment insert test should return true
+    public void insertKnownPost()
+    {
         assertTrue(database.insertKnownPost("123456789012345678901234567890123456", null));
+        // insert a duplicate should fail
+        assertFalse(database.insertKnownPost("123456789012345678901234567890123456", null));
         //empty comment string should return true
-        assertTrue(database.insertKnownPost("123456789012345678901234567890123456", ""));
+        assertTrue(database.insertKnownPost("HASH_2789012345678901234567890123456", ""));
```

Comment:
This line and the line below. Why does the test hash start with 'HASH'?

Reply:
it starts with hash so that they are different strings, I try to make them all different when the need to be but also predictable so that there are not duplicates in different areas